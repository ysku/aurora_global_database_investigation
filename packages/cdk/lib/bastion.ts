import {
  GenericLinuxImage,
  Instance,
  InstanceClass,
  InstanceSize,
  InstanceType,
  SecurityGroup,
  Vpc,
} from '@aws-cdk/aws-ec2';
import { ManagedPolicy, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { App, Stack, StackProps } from '@aws-cdk/core';

export type BastionProps = {
  dbAccessSG: SecurityGroup;
  vpc: Vpc;
} & StackProps;

export class BastionStack extends Stack {
  constructor(scope: App, id: string, props: BastionProps) {
    super(scope, id, props);

    const { dbAccessSG, vpc } = props;

    new Instance(this, 'bastion', {
      instanceName: 'bastion',
      vpc,
      instanceType: InstanceType.of(InstanceClass.T2, InstanceSize.MICRO),
      machineImage: new GenericLinuxImage({
        'ap-northeast-1': 'ami-059b6d3840b03d6dd', // Ubuntu Server 20.04 LTS (HVM), SSD Volume Type
      }),
      securityGroup: dbAccessSG,
      role: new Role(this, 'bastion-role', {
        assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
        managedPolicies: [
          ManagedPolicy.fromAwsManagedPolicyName(
            'AmazonSSMManagedInstanceCore'
          ),
        ],
      }),
    });
  }
}
