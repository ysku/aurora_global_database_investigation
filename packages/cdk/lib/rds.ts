import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  SecurityGroup,
  SubnetType,
  Vpc,
} from '@aws-cdk/aws-ec2';
import {
  AuroraPostgresEngineVersion,
  Credentials,
  DatabaseCluster,
  DatabaseClusterEngine,
  ParameterGroup,
} from '@aws-cdk/aws-rds';
import { Secret } from '@aws-cdk/aws-secretsmanager';
import { App, Stack, StackProps } from '@aws-cdk/core';

export type RDSStackProps = {
  dbClusterIdentifier: string;
  dbName: string;
  dbSG: SecurityGroup;
  dbUserName: string;
  vpc: Vpc;
} & StackProps;

export class RDSStack extends Stack {
  readonly dbCluster: DatabaseCluster;
  readonly secret: Secret;

  constructor(scope: App, id: string, props: RDSStackProps) {
    super(scope, id, props);

    const { dbClusterIdentifier, dbName, dbSG, dbUserName, vpc } = props;

    // use secret manager to configure database username and password
    // cf. https://docs.aws.amazon.com/ja_jp/secretsmanager/latest/userguide/intro.html
    this.secret = new Secret(this, 'Secret', {
      secretName: `${id}Secret`,
      description: 'RDS database auto-generated user password',
      generateSecretString: {
        excludeCharacters: ':/?#[]@!$&"\'()*+,;=%{}|\\^~[]`',
        generateStringKey: 'password',
        passwordLength: 32,
        secretStringTemplate: `{"username": "${dbUserName}"}`,
      },
    });

    const parameterGroup = new ParameterGroup(this, 'RDSParameterGroup', {
      description: `parameter group for ${dbName}`,
      engine: DatabaseClusterEngine.auroraPostgres({
        version: AuroraPostgresEngineVersion.VER_10_7,
      }),
      parameters: {
        log_statement: 'all',
      },
    });

    this.dbCluster = new DatabaseCluster(this, 'RDS', {
      clusterIdentifier: dbClusterIdentifier,
      // cannot use upper case characters.
      defaultDatabaseName: dbName,
      parameterGroup: parameterGroup,
      instances: 2,
      instanceProps: {
        instanceType: InstanceType.of(InstanceClass.R4, InstanceSize.LARGE),
        vpc: vpc,
        vpcSubnets: {
          subnetType: SubnetType.ISOLATED,
        },
        securityGroups: [dbSG],
      },
      port: 5432,
      engine: DatabaseClusterEngine.auroraPostgres({
        version: AuroraPostgresEngineVersion.VER_10_7,
      }),
      credentials: Credentials.fromSecret(this.secret),
    });
  }
}
