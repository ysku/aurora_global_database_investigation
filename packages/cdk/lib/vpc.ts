import {
  CfnSecurityGroupIngress,
  SecurityGroup,
  SubnetType,
  Vpc,
  VpcProps,
} from '@aws-cdk/aws-ec2';
import { App, Stack, StackProps } from '@aws-cdk/core';

export class VPCStack extends Stack {
  readonly vpc: Vpc;
  readonly dbAccessSG: SecurityGroup;
  readonly dbSG: SecurityGroup;
  readonly instanceSG: SecurityGroup;

  constructor(scope: App, id: string, props?: StackProps) {
    super(scope, id, props);

    const vpcProps: VpcProps = {
      cidr: '192.168.0.0/16',
      enableDnsHostnames: true,
      enableDnsSupport: true,
      maxAzs: 3,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public',
          subnetType: SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private',
          subnetType: SubnetType.PRIVATE,
        },
        {
          cidrMask: 24,
          name: 'isolated',
          subnetType: SubnetType.ISOLATED,
        },
      ],
    };

    // NOTE: default security group will be also created
    this.vpc = new Vpc(this, 'VPC', vpcProps);

    this.dbAccessSG = new SecurityGroup(this, 'dbaccesssg', {
      vpc: this.vpc,
      description: 'for instances that access database',
      securityGroupName: 'Database Access',
    });

    this.dbSG = new SecurityGroup(this, 'dbsg', {
      vpc: this.vpc,
      description: 'for database',
      securityGroupName: 'Database',
    });

    this.instanceSG = new SecurityGroup(this, 'instancesg', {
      vpc: this.vpc,
      description: 'for ec2 instance',
      securityGroupName: 'Instance',
    });

    // allow DatabaseSG to access from DBAccessSG
    new CfnSecurityGroupIngress(this, 'fromdbaccesssg', {
      groupId: this.dbSG.securityGroupId,
      ipProtocol: 'tcp',
      description: `from instances with sg named ${this.dbAccessSG.securityGroupName}`,
      fromPort: 5432,
      toPort: 5432,
      sourceSecurityGroupId: this.dbAccessSG.securityGroupId,
    });

    // allow DatabaseSG to access from DBAccessSG
    new CfnSecurityGroupIngress(this, 'frominstance', {
      groupId: this.dbSG.securityGroupId,
      ipProtocol: 'tcp',
      description: `from instances with sg named ${this.instanceSG.securityGroupName}`,
      fromPort: 5432,
      toPort: 5432,
      sourceSecurityGroupId: this.instanceSG.securityGroupId,
    });
  }
}
