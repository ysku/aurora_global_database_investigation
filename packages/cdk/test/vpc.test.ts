import { SynthUtils } from '@aws-cdk/assert';
import { App } from '@aws-cdk/core';

import { VPCStack } from '../lib/vpc';

describe('VPCStack', () => {
  const app = new App();

  test('snapshot', () => {
    const stack = new VPCStack(app, 'TestVPCStack');
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
