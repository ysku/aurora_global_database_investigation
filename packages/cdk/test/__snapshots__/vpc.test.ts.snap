// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VPCStack snapshot 1`] = `
Object {
  "Resources": Object {
    "VPCB9E5F0B4": Object {
      "Properties": Object {
        "CidrBlock": "192.168.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VPCIGWB7E252D3": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VPCVPCGW99B986DC": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "VPCisolatedSubnet1RouteTable2BB88525": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/isolatedSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCisolatedSubnet1RouteTableAssociation2886DD39": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCisolatedSubnet1RouteTable2BB88525",
        },
        "SubnetId": Object {
          "Ref": "VPCisolatedSubnet1Subnet84AF7038",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCisolatedSubnet1Subnet84AF7038": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "192.168.4.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "isolated",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/isolatedSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCisolatedSubnet2RouteTable1EDFC77F": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/isolatedSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCisolatedSubnet2RouteTableAssociation8A6ECC3C": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCisolatedSubnet2RouteTable1EDFC77F",
        },
        "SubnetId": Object {
          "Ref": "VPCisolatedSubnet2Subnet8AE9E765",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCisolatedSubnet2Subnet8AE9E765": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "192.168.5.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "isolated",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/isolatedSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCprivateSubnet1DefaultRoute68B79172": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VPCpublicSubnet1NATGatewayEEDEB5A1",
        },
        "RouteTableId": Object {
          "Ref": "VPCprivateSubnet1RouteTable41566AF9",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCprivateSubnet1RouteTable41566AF9": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/privateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCprivateSubnet1RouteTableAssociation43FE0C59": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCprivateSubnet1RouteTable41566AF9",
        },
        "SubnetId": Object {
          "Ref": "VPCprivateSubnet1SubnetE522290D",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCprivateSubnet1SubnetE522290D": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "192.168.2.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/privateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCprivateSubnet2DefaultRouteD808B206": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VPCpublicSubnet2NATGateway13FC4601",
        },
        "RouteTableId": Object {
          "Ref": "VPCprivateSubnet2RouteTableBBE8BB2B",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCprivateSubnet2RouteTableAssociation45BBC356": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCprivateSubnet2RouteTableBBE8BB2B",
        },
        "SubnetId": Object {
          "Ref": "VPCprivateSubnet2Subnet2F4D9530",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCprivateSubnet2RouteTableBBE8BB2B": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/privateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCprivateSubnet2Subnet2F4D9530": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "192.168.3.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/privateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCpublicSubnet1DefaultRoute19059ECD": Object {
      "DependsOn": Array [
        "VPCVPCGW99B986DC",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "RouteTableId": Object {
          "Ref": "VPCpublicSubnet1RouteTableF591E248",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCpublicSubnet1EIP2D50A06D": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/publicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VPCpublicSubnet1NATGatewayEEDEB5A1": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "VPCpublicSubnet1EIP2D50A06D",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "VPCpublicSubnet1Subnet325F50B2",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/publicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VPCpublicSubnet1RouteTableAssociationBFFA43B7": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCpublicSubnet1RouteTableF591E248",
        },
        "SubnetId": Object {
          "Ref": "VPCpublicSubnet1Subnet325F50B2",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCpublicSubnet1RouteTableF591E248": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/publicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCpublicSubnet1Subnet325F50B2": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "192.168.0.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/publicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCpublicSubnet2DefaultRoute9ED4D6BC": Object {
      "DependsOn": Array [
        "VPCVPCGW99B986DC",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "RouteTableId": Object {
          "Ref": "VPCpublicSubnet2RouteTable55DAB1BB",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCpublicSubnet2EIPB6C99FBD": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/publicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VPCpublicSubnet2NATGateway13FC4601": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "VPCpublicSubnet2EIPB6C99FBD",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "VPCpublicSubnet2Subnet2B2DFF71",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/publicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VPCpublicSubnet2RouteTable55DAB1BB": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/publicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCpublicSubnet2RouteTableAssociation0849466B": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCpublicSubnet2RouteTable55DAB1BB",
        },
        "SubnetId": Object {
          "Ref": "VPCpublicSubnet2Subnet2B2DFF71",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCpublicSubnet2Subnet2B2DFF71": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "192.168.1.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "TestVPCStack/VPC/publicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "dbaccesssg59FD6FC1": Object {
      "Properties": Object {
        "GroupDescription": "for instances that access database",
        "GroupName": "Database Access",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "dbsgB988D619": Object {
      "Properties": Object {
        "GroupDescription": "for database",
        "GroupName": "Database",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "fromdbaccesssg": Object {
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "from instances with sg named ",
              Object {
                "Ref": "dbaccesssg59FD6FC1",
              },
            ],
          ],
        },
        "FromPort": 5432,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "dbsgB988D619",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "dbaccesssg59FD6FC1",
            "GroupId",
          ],
        },
        "ToPort": 5432,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "frominstance": Object {
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "from instances with sg named ",
              Object {
                "Ref": "instancesg9E99325E",
              },
            ],
          ],
        },
        "FromPort": 5432,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "dbsgB988D619",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "instancesg9E99325E",
            "GroupId",
          ],
        },
        "ToPort": 5432,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "instancesg9E99325E": Object {
      "Properties": Object {
        "GroupDescription": "for ec2 instance",
        "GroupName": "Instance",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
}
`;
